### PR TITLE
Add Vocab to Python module.

### PR DESCRIPTION
**Pull request by sebpuetz**
Would that fit what you described in #5?

Can be used in python as:

```Python
import finalfusion
import numpy as np

embeddings = finalfusion.Embeddings("embeddings.fifu")
vocab = embeddings.vocab()
word_0 = vocab[0]
last_word = vocab[-1]
word2idx = dict([(word, idx) for idx, word in enumerate(vocab)])

idx = vocab.item_to_indices("InVocab")
in_vocab_embed = matrix[idx]

indices = vocab.item_to_indices("OOV")
matrix = embeddings.matrix_copy()
oov_embed = np.zeros(300)
for index in indices:
    oov_embed += matrix[index]
```

```Python
import finalfusion

embeddings = finalfusion.Embeddings("embeddings.fifu")
vocab = embeddings.vocab()
vocab[len(vocab)] # raises exception
```

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -309,6 +318,64 @@ impl PyIterProtocol for PyEmbeddings {
...
+    embeddings: Rc<RefCell<EmbeddingsWrap>>,
+}
+
+#[pymethods]
+impl PyVocab {
+    fn item_to_indices(&self, key: String) -> PyResult<PyObject> {
+        let embeds = self.embeddings.borrow();
+
+        use EmbeddingsWrap::*;
+        let indices = match &*embeds {
```
Comment (danieldk) [inline src/lib.rs:333]: This can be a separate PR: but I am wondering if we shouldn't just implement a basic set of methods on `EmbeddingsWrap`, then we don't need all the `match`ing in the Python method implementations.

File: src/lib.rs [hunk 2]
```diff
@@ -309,6 +318,64 @@ impl PyIterProtocol for PyEmbeddings {
...
+            View(e) => e.vocab().idx(key.as_str()),
+            NonView(e) => e.vocab().idx(key.as_str()),
+        };
+        if let Some(indices) = indices {
```
Comment (danieldk) [inline src/lib.rs]: If both cases are handled, I have a preference for `match`. You may also want to try and see how something along the lines of:

```
indices.map(|indices| { ... }).ok_or_else(|| Err(exceptions::KeyError::py_err("Unknown word or n-grams"))
```
looks aesthetically.

File: src/lib.rs [hunk 3]
```diff
@@ -309,6 +318,64 @@ impl PyIterProtocol for PyEmbeddings {
...
+        let embeds = self.embeddings.borrow();
+
+        use EmbeddingsWrap::*;
+
+        let v = match &*embeds {
+            View(e) => e.vocab(),
+            NonView(e) => e.vocab(),
+        };
+
+        if (v.words().len() as isize) <= idx || idx < 0 {
```
Comment (danieldk) [inline src/lib.rs]: Isn't `idx >= v.words.len() as isize` more readable? Also, `as` has high precedence, so the parentheses are not needed. Nowadays you could also replace the condition by:

`(0..v.word.len()).contains(&idx)`

But I didn't check which Rust version added this (when later than 1.31.0, we can't use it yet).

**Review by danieldk — CHANGES_REQUESTED** [review]
Two small style comments.

Looks good otherwise!

File: src/lib.rs (hunk 1 above)
Comment (sebpuetz) [reply src/lib.rs:333]: I was thinking the same. Getting the VocabWrap could be implemented, and maybe some methods that are shared among the different Storage implementations.

It's not possible to implement Deref for EmbeddingsWrap because of the type parameters, right?

File: src/lib.rs (hunk 3 above)
Comment (sebpuetz) [reply src/lib.rs]: > Isn't idx >= v.words.len() as isize more readable? 

Sure!

>Also, as has high precedence, so the parentheses are not needed. 

Fixed.

>But I didn't check which Rust version added this (when later than 1.31.0, we can't use it yet).

Pretty sure `Range::contains` was added in a very recent version. 1.35 or even 1.36.

File: src/lib.rs (hunk 2 above)
Comment (sebpuetz) [reply src/lib.rs]: Done

File: src/lib.rs (hunk 1 above)
Comment (danieldk) [reply src/lib.rs:333]: > It's not possible to implement Deref for EmbeddingsWrap because of the type parameters, right?

Right, but this would go a long way in simplifying the API:

~~~rust
impl EmbeddingsWrap {
    fn storage(&self) -> &Storage {
        use EmbeddingsWrap::*;

        match self {
            View(e) => e.storage(),
            NonView(e) => e.storage(),
        }
    }

    fn vocab(&self) -> &VocabWrap {
        use EmbeddingsWrap::*;

        match self {
            View(e) => e.vocab(),
            NonView(e) => e.vocab(),
        }
    }

}
~~~

**Review by danieldk — APPROVED** [review]
Looks good! Will merge when CI is happy.
